[RA2 Ch04] - Container runtime configs. (Issue 1636) (#1995)
cntt-n/CNTT: doc/ref_arch/kubernetes/chapters/chapter04.md

|`ra2.crt.005`|[User](https://github.com/opencontainers/runtime-spec/blob/master/config.md#user)|User for the process is a platform-specific structure that allows specific control over which user the process runs as|TBD|
|`ra2.crt.006`|[Platform-specific configuration](https://github.com/opencontainers/runtime-spec/blob/master/config.md#platform-specific-configuration)|This specifies the set of platform (linux/windows,solaris,vm) specific configurations|TBD|
|`ra2.crt.007`|[POSIX-platform Hooks](https://github.com/opencontainers/runtime-spec/blob/master/config.md#posix-platform-hooks)|For POSIX platforms, the configuration structure supports hooks for configuring custom actions related to the lifecycle of the container|TBD|
|`ra2.crt.008`|[Annotations](https://github.com/opencontainers/runtime-spec/blob/master/config.md#annotations)|Contains arbitrary metadata for the container|TBD|
|`ra2.crt.009`|[Extensibility](https://github.com/opencontainers/runtime-spec/blob/master/config.md#extensibility)|Runtimes MAY log unknown properties but MUST otherwise ignore them|TBD|

> |`ra2.crt.0010`|[Valid values](https://github.com/opencontainers/runtime-spec/blob/master/config.md#valid-values)|Runtimes MUST generate an error when invalid or unsupported values are encountered|TBD|

Reviewer: Having read through again a few times, I'm not sure what value listing these brings.  If we're stating that the container runtime must conform to OCI 1.0 then all of these should be assumed?  i.e. it would not be possible for a runtime to conform to OCI1.0 and comply with some of these config groups, but not others?

I think this section of ch4 would be more useful for RI/VI/RC if it were to include the configuration specifics we need a container runtime to be able to support/specify, e.g. "the container runtime must run as a system daemon / service" or "the container runtime daemon must run with IPv6 configuration options enabled" (to support IPv6 requirements, for example).  Or specifications relating to running multiple runtimes, or specifications to address requirements around isolation (e.g. kata/gvisor), etc.

Hope that makes sense?

Author: Are u suggesting single entry in table here to specify that runtimes should be compliant with OCI 1.0? Details of each parameter group should go RI/VI/RC?

Reviewer: No. I'm suggesting this table should include the configuration of a runtime or runtimes needed to meet the requirements.  The config.md seems to me to be about how a runtime deals with a container bundle, and what the config.json part of the container bundle needs to include - I feel that is useful for the developers of the container runtime (see [here](https://github.com/opencontainers/runtime-spec#use-cases)) so I think they'd be better discussed as part of section 4.9 in the next release, in terms of providing input to application developers to ensure their containers are compliant with the OCI specs. 

For example, if we included them all, we would need to create test cases in RC2 to test that "any" runtime or runtime engine (e.g. containerd) did indeed generate an error when invalid or unsupported values are encountered in the container config.json ([Valid values](https://github.com/opencontainers/runtime-spec/blob/master/config.md#valid-values)).  In effect we'd be testing the compliance of the runtime to the OCI spec, which is fine if we can simply call an existing conformance test suite. Unfortunately there isn't such a thing for OCI runtime spec today (only the [distribution spec](https://github.com/opencontainers/oci-conformance/tree/master/distribution-spec)).  From reading around, most projects/vendors make use of runc or another OCI-compliance runtime as part of their bigger "engine" (containerd, Docker, CRI-O, etc.) in order to claim compliance.

So for this table, I think we should focus on the choices that an implementation needs to make when it comes to the container runtime when it is being installed and is running (i.e. what flags need using, and with which values).

Author: I understood your concern. But that is a huge list of parameters and their possible values. Which is impossible to put here in tabular format. That's why i have put the parameter group here along with the link to OCI 1.0 section. Even parameter values differ with platform type too.
Sorry but could not find a better solution for this. Pls  let me know any other alternative if you have.

Reviewer: > Even parameter values differ with platform type too.

Exactly - but those parameter values are useful for the workload developers, not for the deployment of a runtime into an operating system which is what this table should focus on.  I think move these to 4.9 please (change the ref accordingly) and then we can separately look to include more in this section.

Author: Makes sense. I have updated PR. Pls check now.

> <p align="center"><b>Table 4-3:</b> OCI 1.0 Conferment Container Runtime Configurations</p>

Reviewer: ```suggestion
<p align="center"><b>Table 4-3:</b> Container Runtime Specifications</p>
```

## 4.9 Kubernetes workloads

In order for the Kubernetes workloads to be conformant with the Reference

> Architecture they must implement the parameter values as per the following specifications:

Reviewer: Mention the parameter groups in the spec, as there will be other specifications not related to parameter groups to come in the next release.
```suggestion
Architecture they must be implemented as per the following specifications:
```

> |Ref|Parameter Group|Details|Requirement Trace|

Reviewer: Mention the parameter groups in the specs themselves, as there will be other specifications not related to parameter groups to come in the next release.
```suggestion
|Ref|Specification|Details|Requirement Trace|
```

|---|---|---|---|

> |`ra2.app.001`|[Root](https://github.com/opencontainers/runtime-spec/blob/master/config.md)|Specifies the container's root filesystem.|TBD|
> |`ra2.app.002`|[Mounts](https://github.com/opencontainers/runtime-spec/blob/master/config.md#mounts)|Specifies additional mounts beyond root|TBD|
> |`ra2.app.003`|[Process](https://github.com/opencontainers/runtime-spec/blob/master/config.md#process)|Specifies the container process|TBD|
> |`ra2.app.004`|[Hostname](https://github.com/opencontainers/runtime-spec/blob/master/config.md#hostname)|Specifies the container's hostname as seen by processes running inside the container|TBD|
> |`ra2.app.005`|[User](https://github.com/opencontainers/runtime-spec/blob/master/config.md#user)|User for the process is a platform-specific structure that allows specific control over which user the process runs as|TBD|

Reviewer: Mention the parameter groups in the spec, as there will be other specifications not related to parameter groups to come in the next release.
```suggestion
|`ra2.app.001`|[Root](https://github.com/opencontainers/runtime-spec/blob/master/config.md) Parameter Group (OCI Spec)|Specifies the container's root filesystem.|TBD|
|`ra2.app.002`|[Mounts](https://github.com/opencontainers/runtime-spec/blob/master/config.md#mounts) Parameter Group (OCI Spec)|Specifies additional mounts beyond root|TBD|
|`ra2.app.003`|[Process](https://github.com/opencontainers/runtime-spec/blob/master/config.md#process) Parameter Group (OCI Spec)|Specifies the container process|TBD|
|`ra2.app.004`|[Hostname](https://github.com/opencontainers/runtime-spec/blob/master/config.md#hostname) Parameter Group (OCI Spec)|Specifies the container's hostname as seen by processes running inside the container|TBD|
|`ra2.app.005`|[User](https://github.com/opencontainers/runtime-spec/blob/master/config.md#user) Parameter Group (OCI Spec)|User for the process is a platform-specific structure that allows specific control over which user the process runs as|TBD|
```

> |`ra2.app.006`|[Platform-specific configuration](https://github.com/opencontainers/runtime-spec/blob/master/config.md#platform-specific-configuration)|This specifies the set of platform (linux/windows,solaris,vm) specific configurations|TBD|
> |`ra2.app.007`|[POSIX-platform Hooks](https://github.com/opencontainers/runtime-spec/blob/master/config.md#posix-platform-hooks)|For POSIX platforms, the configuration structure supports hooks for configuring custom actions related to the lifecycle of the container|TBD|
> |`ra2.app.008`|[Annotations](https://github.com/opencontainers/runtime-spec/blob/master/config.md#annotations)|Contains arbitrary metadata for the container|TBD|
> |`ra2.app.009`|[Extensibility](https://github.com/opencontainers/runtime-spec/blob/master/config.md#extensibility)|Runtimes MAY log unknown properties but MUST otherwise ignore them|TBD|
> |`ra2.app.0010`|[Valid values](https://github.com/opencontainers/runtime-spec/blob/master/config.md#valid-values)|Runtimes MUST generate an error when invalid or unsupported values are encountered|TBD|

Reviewer: Suggest these ones are removed as their not about the apps.
```suggestion
```

Author: Done
